src/main/java/hudson/plugins/gradle/injection/MavenFreeStyleRunListener.java

```diff
@@ -0,0 +1,18 @@
+package hudson.plugins.gradle.injection;
+
+import hudson.Extension;
+import hudson.model.FreeStyleBuild;
+import hudson.model.TaskListener;
+import hudson.model.listeners.RunListener;
+
+@Extension
+public class MavenFreeStyleRunListener extends RunListener<FreeStyleBuild> implements MavenInjectionAware {
+
+    @Override
+    public void onStarted(FreeStyleBuild freeStyleBuild, TaskListener listener) {
```

Comment:
Is it guaranteed that it's called only once?

Reply:
The documentation doesn't mentioned it, but by looking at the code and logs of job execution, I didn't find instances of it being called more than once. 
We can add a check for an existing `NodePreparedMavenEnvsAction` object in `Run` instance and not add it if exists.

Reply:
I added a check to verify Action object is not in the run execution